src/main/java/meteordevelopment/meteorclient/utils/misc/PooledList.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * This file is part of the Meteor Client distribution (https://github.com/MeteorDevelopment/meteor-client).
+ * Copyright (c) Meteor Development.
+ */
+
+package meteordevelopment.meteorclient.utils.misc;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+public class PooledList<T> implements Iterable<T> {
+    protected final Pool<T> pool;
+    protected final List<T> list = new ArrayList<>();
+
+    public PooledList(Producer<T> producer) {
+        pool = new Pool<>(producer);
+    }
+
+    public T get() {
+        T obj = pool.get();
+        list.add(obj);
+        return obj;
+    }
+
+    public T peekList() {
```

Comment:
Can you rename this method to `getFirst()` to be more clear. Also make it return null if the list is empty since right now it will throw an exception. Also while you are at it, why not add a `getLast()` method too.

Reply:
Problem with that is the other `get` method removes an item from the pool and adds it to the list, while this one peeks the top of the list only. I think it'd be clearer to rename `get` to something else

Reply:
Well yeah, I also made a comment for that.